File: transpile/javascript/swing/AbstractButton.js
/**
 * The javax.swing.AbstractButton clone
 *
 * @author gianpiero.diblasi
 */
class AbstractButton extends JSComponent {

   listeners = new Array();

  /**
   * Creates the object
   *
   * @param element The HTML element representing this component
   */
  constructor(element) {
    super(element);
  }

  /**
   * Clone of javax.swing.AbstractButton.addActionListener
   *
   * @param listener The listener
   */
   addActionListener(listener) {
    this.listeners.push(listener);
  }

  /**
   * The method for click events
   */
   onclick() {
    let event = new ActionEvent();
    this.listeners.forEach(listener => {
      if (typeof listener === "function") {
        listener(event);
      } else {
        listener.actionPerformed(event);
      }
    });
  }

  /**
   * Clone of javax.swing.AbstractButton.setTooltip
   *
   * @param text The text
   */
   setTooltip(text) {
    this.setAttribute("title", text);
  }

  /**
   * Sets the icon
   *
   * @param producer The icon producer
   */
   setIcon(producer) {
    this.prependNodeChild(producer.produce());
  }
}

File: deposito/AbstractButton.java
package javascript.swing;

import def.js.Array;
import javascript.awt.event.ActionEvent;
import javascript.awt.event.ActionListener;
import static simulation.js.$Globals.$typeof;

/**
 * The javax.swing.AbstractButton clone
 *
 * @author gianpiero.diblasi
 */
public abstract class AbstractButton extends JSComponent {

  private final Array<ActionListener> listeners = new Array<>();

  public AbstractButton() {
    super();
  }

  /**
   * Clone of javax.swing.AbstractButton.addActionListener
   * @param listener The listener
   */
  public void addActionListener(ActionListener listener) {
    this.listeners.push(listener);
  }

  /**
   * The method for click events
   *
   * @return null
   */
  protected Object onclick() {
    ActionEvent event = new ActionEvent();

    this.listeners.forEach(listener -> {
      if ($typeof(listener, "function")) {
        listener.$apply(event);
      } else {
        listener.actionPerformed(event);
      }
    });
    return null;
  }
}
